src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -532,19 +550,21 @@ public FSFileInfo linkSelfHostedFile(
                     true,
                     new FSItem()
                         .path(fileLoadModel.getTargetPath())
-                        .checksumMd5(sourceBlob.getMd5ToHexString())
+                        .checksumMd5(finalMd5.effectiveMd5())
                         .size(sourceBlob.getSize()))
                 .toString();
       } else {
         effectiveFileId = fileId;
       }
       // MD5 is computed per-component. So if there are multiple components, the MD5 here is
       // not useful for validating the contents of the file on access. Therefore, we only
-      // return the MD5 if there is only a single component. For more details,
-      // see https://cloud.google.com/storage/docs/hashes-etags
+      // return the MD5 if there is only a single component or if it's been specified by the user.
+      // For more details, see https://cloud.google.com/storage/docs/hashes-etags
       Integer componentCount = sourceBlob.getComponentCount();
       String checksumMd5 = null;
-      if (componentCount == null || componentCount == 1) {
+      if (finalMd5.type().equals(Md5Type.USER_PROVIDED)) {
+        checksumMd5 = finalMd5.effectiveMd5();
+      } else if (componentCount == null || componentCount == 1) {
         checksumMd5 = sourceBlob.getMd5ToHexString();
       }
```

Comment:
Would this block be eligible to be broken out for reuse?  It looks like this duplicates logic from 492 - 502, and that logic has grown a bit in complexity.

Reply:
Fair enough.  Let me combine